### PR TITLE
Add support for L3 PCBs

### PR DESCRIPTION
**Pull request by MajorKoos**
## Add support for L3 PCBs
 - L3 TKL
 - LZ CLSm
 - LZ Ergo
 - LZ St

## Notes
- These PCBs don't like anything > 10.54. (repeated keystrokes and hangs).  
- It seems related to the additional HID endpoints which were added in 0.11.x.
- Setting keyboard_shared_ep=yes *fixes* it
- All the LEDs are controlled by an atiny85 with it's own set of effects baked in.  
- I've disabled those and let the QMK run them off the atmega32a rather than offloading to the atiny85 like JigOn does.  Pick your poison I guess.
- I've disabled RGB and backlighting on the LZ Ergo entirely because of known issues with the PCB.
- It can be re-enabled by those brave souls with an ISP if something goes wrong.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/leeku/l3_tkl/l3_tkl_via.json [hunk 1]
```diff
@@ -0,0 +1,375 @@
+{
```
Comment (zvecr) [inline keyboards/leeku/l3_tkl/l3_tkl_via.json:1]: Should not be in this repo

File: keyboards/leeku/l3_tkl/l3_tkl_layout.json [hunk 2]
```diff
@@ -0,0 +1,399 @@
+{
```
Comment (zvecr) [inline keyboards/leeku/l3_tkl/l3_tkl_layout.json:1]: remove

File: keyboards/leeku/l3_tkl/keymaps/via/rules.mk [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+VIA_ENABLE = yes            # VIA support should be enabled here due to the main() loop will be compiled first.
+BOOTMAGIC_ENABLE = no      # Virtual DIP switch configuration
```
Comment (zvecr) [inline keyboards/leeku/l3_tkl/keymaps/via/rules.mk:17]:
```suggestion
```

File: keyboards/leeku/i2c.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
+/*
```
Comment (zvecr) [inline keyboards/leeku/i2c.c:1]: Should not introduce its own i2c code, and use <https://docs.qmk.fm/#/i2c_driver>

File: keyboards/leeku/i2c.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
+/*
```
Comment (zvecr) [inline keyboards/leeku/i2c.h:1]: same as .c file

File: keyboards/leeku/hwaddress.h [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef HWADDRESS_H
+#define HWADDRESS_H
+
```
Comment (zvecr) [inline keyboards/leeku/hwaddress.h:19]:
```suggestion
#pragma once

```

File: keyboards/leeku/hwaddress.h [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    uint8_t rgb_chain;                      // RGB5050 numbers (H/W dependent)
+    rgb_effect_param_type rgb_effect_param[RGB_EFFECT_MAX]; // RGB effect parameter
+}kbd_configuration_t;
+
+#define MACRO_ADDR_START      0x4400     // 0x4400 ~ 0x6FFF  (7KBytes - 256B x 44)
+
+extern kbd_configuration_t kbdConf;
+extern void updateConf(void);
+
+#endif // HWADDRESS_H
```
Comment (zvecr) [inline keyboards/leeku/hwaddress.h:90]:
```suggestion
```

File: keyboards/leeku/l3_tkl/keymaps/default/keymap.c [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include QMK_KEYBOARD_H
+
+#define ______ KC_TRNS
```
Comment (zvecr) [inline keyboards/leeku/l3_tkl/keymaps/default/keymap.c:18]: Should use the built in `_______`
```suggestion
```

File: keyboards/leeku/l3_tkl/l3_tkl.h [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "quantum.h"
+
+#define _NO KC_NO
+
+#define LAYOUT_TKL( \
```
Comment (zvecr) [inline keyboards/leeku/l3_tkl/l3_tkl.h:23]: required to be `LAYOUT_` and then lowercase with underscores

```suggestion
#define LAYOUT_tkl( \
```

File: keyboards/leeku/led_l3.h [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef L3LED_H
+#define L3LED_H
```
Comment (zvecr) [inline keyboards/leeku/led_l3.h:19]:
```suggestion
#pragma once
```

File: keyboards/leeku/led_l3.h [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+void led_sleep(void);
+void led_restore(void);
+void led_rgb_preset(void);
+void led_change_mode(void);
+void rgb_change_mode(void);
+void rgb_set_mode(uint8_t mode);
+void rgb_set_all(uint8_t on, uint8_t r, uint8_t g, uint8_t b);
+void rgblight_set(void);
+
+#endif // L3LED_H
```
Comment (zvecr) [inline keyboards/leeku/led_l3.h:53]:
```suggestion
```

File: keyboards/leeku/lz_clsm/keymaps/default/keymap.c [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "../via/keymap.c"
```
Comment (zvecr) [inline keyboards/leeku/lz_clsm/keymaps/default/keymap.c:16]:
```suggestion
```

should have a full keymap

File: keyboards/leeku/lz_clsm/keymaps/via/keymap.c [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include QMK_KEYBOARD_H
+#include "stdint.h"
+
+#define ______ KC_TRNS
```
Comment (zvecr) [inline keyboards/leeku/lz_clsm/keymaps/via/keymap.c:19]:
```suggestion

```

**Review by zvecr — CHANGES_REQUESTED** [review]
A **_lot_** of files suffer from the same issues, so im not going to flag every case (part of the issue with PRing multiple boards at the same time).

File: keyboards/leeku/l3_tkl/l3_tkl_via.json (hunk 1 above)
Comment (MajorKoos) [reply keyboards/leeku/l3_tkl/l3_tkl_via.json]: Removed file

File: keyboards/leeku/l3_tkl/l3_tkl_layout.json (hunk 2 above)
Comment (MajorKoos) [reply keyboards/leeku/l3_tkl/l3_tkl_layout.json]: Removed file

File: keyboards/leeku/l3_tkl/keymaps/via/rules.mk (hunk 3 above)
Comment (MajorKoos) [reply keyboards/leeku/l3_tkl/keymaps/via/rules.mk]: Removed redundant declaration

File: keyboards/leeku/i2c.c (hunk 4 above)
Comment (MajorKoos) [reply keyboards/leeku/i2c.c]: Replaced custom I2C driver with QMK I2C driver

File: keyboards/leeku/i2c.h (hunk 5 above)
Comment (MajorKoos) [reply keyboards/leeku/i2c.h]: Replaced custom I2C driver with QMK I2C driver

File: keyboards/leeku/l3_tkl/l3_tkl.h (hunk 9 above)
Comment (MajorKoos) [reply keyboards/leeku/l3_tkl/l3_tkl.h]: Done.

File: keyboards/leeku/led_l3.h (hunk 10 above)
Comment (MajorKoos) [reply keyboards/leeku/led_l3.h]: Done.

File: keyboards/leeku/led_l3.h (hunk 11 above)
Comment (MajorKoos) [reply keyboards/leeku/led_l3.h]: Done.

File: keyboards/leeku/lz_clsm/keymaps/default/keymap.c (hunk 12 above)
Comment (MajorKoos) [reply keyboards/leeku/lz_clsm/keymaps/default/keymap.c]: Included the full keymap for default

File: keyboards/leeku/lz_clsm/keymaps/via/keymap.c (hunk 13 above)
Comment (MajorKoos) [reply keyboards/leeku/lz_clsm/keymaps/via/keymap.c]: replaced with built in macro

File: keyboards/leeku/hwaddress.h (hunk 6 above)
Comment (MajorKoos) [reply keyboards/leeku/hwaddress.h]: Done.

File: keyboards/leeku/l3_tkl/keymaps/default/keymap.c (hunk 8 above)
Comment (MajorKoos) [reply keyboards/leeku/l3_tkl/keymaps/default/keymap.c]: Done.

File: keyboards/leeku/hwaddress.h (hunk 7 above)
Comment (MajorKoos) [reply keyboards/leeku/hwaddress.h]: Done.

**Comment by MajorKoos** [comment]
Thanks for the feedback. I've made the requested changes.

**Review by MajorKoos — COMMENTED** [review]
Thanks, I've pulled out the old I2C driver and updated the L3 API to better match QMK standards.

File: keyboards/leeku/l3_tkl/config.h [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include "config_common.h"
+
+/* USB Device descriptor parameter */
+#define VENDOR_ID       0x20A0
```
Comment (drashna) [inline keyboards/leeku/l3_tkl/config.h:22]: This vendor ID is in use.  Please change it. 

https://www.the-sz.com/products/usbid/index.php?v=0x20A0&p=&n=

**Review by drashna — APPROVED** [review]
So this is using a custom implementation of rgb support, rather than using the normal i2c based rgb light driver?

File: keyboards/leeku/l3_tkl/config.h (hunk 14 above)
Comment (MajorKoos) [reply keyboards/leeku/l3_tkl/config.h:22]: How about FFFE ?
That seems completely unused.
I wasn't really sure what to do about the vendor ID so I picked something that wasn't taken under clay logic.

**Comment by MajorKoos** [comment]
> So this is using a custom implementation of rgb support, rather than using the normal i2c based rgb light driver?

Yeah.  All the bling is offloaded to an attiny85 on these PCBs.
https://github.com/leeku11/l3/blob/master/source/timymain/i2cReceiver/main.c

**Comment by drashna** [comment]
So all the handling is done on the attiny, then?  

If not, then both the rgblight (underglow) and rgb matrix (per key) features support custom drivers, so you coul send the LED data to the attiny's. 

If so, then that's kind of a shame, though, I've seen attiny firmware that just handles sending the LED info.

**Comment by drashna** [comment]
Just a heads up, instead of using the rgblight custom driver, instead you can use `rgblight_call_driver`.   This allows rgblght to do all the stuff.  That allows rgblight to do all the stuff (clipping, splits, the "layer" feature, etc), and this function just handles sending the LED array to the controller, instead.   You can look at the ergodox EZ for reference, as it sends half of the array over i2c to what I think is an attiny, actually.

**Comment by MajorKoos** [comment]
> Just a heads up, instead of using the rgblight custom driver, instead you can use `rgblight_call_driver`. This allows rgblght to do all the stuff. That allows rgblight to do all the stuff (clipping, splits, the "layer" feature, etc), and this function just handles sending the LED array to the controller, instead. You can look at the ergodox EZ for reference, as it sends half of the array over i2c to what I think is an attiny, actually.

Thanks for the guidance.  I'll go that route.

**Comment by MajorKoos** [comment]
I replaced the custom RGB driver with a hook into rgblight_call_driver and then send the data over i2c to the attiny85 in the format it expects.

**Review by drashna — APPROVED** [review]

**Pull request closed without merging**